compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/GPUTileSwizzleUtils.cpp

```diff
@@ -140,8 +145,132 @@ static int getInnermostNonInternalDimIdx(
+  TileSwizzle::ExpandShapeType expandShape = swizzle.expandShape;
+  TileSwizzle::ExpandShapeType accExpandShape = accSwizzle.expandShape;
+  // We will construct the permutation on the flattened `expandShape` dims that
+  // will move the cross thread dims to outermost, and store the resulting
+  // permutation in `crossThreadToOuterPerm`. This will be used later to adjust
+  // the swizzle.permutation properly.
+  SmallVector<int64_t> crossThreadToOuterPerm;
+  int groupStartIdx = 0;
+  for (int accGroupIdx = 0; accGroupIdx < accExpandShape.size();
+       ++accGroupIdx) {
+    TileSwizzle::ExpandShapeDimVectorType accGroup =
+        accExpandShape[accGroupIdx];
+    // Index for corresponding `swizzle` group is 0 for Lhs and Rhs fragments,
+    // since the 1 index of Rhs and Lhs are K dimensions.
+    int groupIdx = fragment == MMAFragment::Acc ? accGroupIdx : 0;
+    // Skip N group for Lhs.
+    if (fragment == MMAFragment::Lhs && accGroupIdx == 1) {
+      continue;
+    }
+    // Skip M group for Rhs.
+    if (fragment == MMAFragment::Rhs && accGroupIdx == 0) {
+      continue;
+    }
+    TileSwizzle::ExpandShapeDimVectorType group = expandShape[groupIdx];
+    // The expanded shape of the `accSwizzle` group may not necessarily match
+    // the expanded shape of the `swizzle` group, so we need to expand or group
+    // the dimensions of `group` so that there exists a corresponding dimension
+    // in `group` for every CrossThread dimension in `accGroup`.
+    if (accGroup.size() != group.size()) {
+      SmallVector<int64_t> accGroupShape =
+          llvm::map_to_vector(accGroup, [](TileSwizzle::Dim d) {
+            return static_cast<int64_t>(d.size);
+          });
+      SmallVector<int64_t> groupShape =
+          llvm::map_to_vector(group, [](TileSwizzle::Dim d) {
+            return static_cast<int64_t>(d.size);
+          });
+      std::optional<SmallVector<ReassociationIndices>> groupReassociation =
+          getReassociationIndicesForCollapse(accGroupShape, groupShape);
+      // For the current MFMA layouts, there should always be a reassociation
+      // found, since the ACC layout is always an expanded form of the combined
+      // LHS and RHS layouts.
+      assert(groupReassociation.has_value() &&
+             "expected to find reassociation");
+      TileSwizzle::ExpandShapeDimVectorType expandedGroup;
+      for (auto [i, reassociation] : llvm::enumerate(*groupReassociation)) {
+        for (int64_t reInd : reassociation) {
+          expandedGroup.push_back(
+              TileSwizzle::Dim(group[i].kind, accGroup[reInd].size));
+        }
+        int expandedPermIdx;
+        for (auto [permIdx, permDim] : llvm::enumerate(swizzle.permutation)) {
+          if (permDim > i) {
+            permDim += reassociation.size() - 1;
+          }
+          if (permDim == i) {
+            expandedPermIdx = permIdx;
+          }
+        }
+        for (int j = 0; j < reassociation.size() - 1; ++j) {
```

Comment:
We should either rewrite it to `j + 1 < reassociation.size()` or follow LLVM loop-style `for (int j = 0, e = reassociation.size() - 1; j < e;`. Because this style of code is problematic when the vector is empty. It becomes an infinite loop. I understand that it is not valid in the reassociation case, but the style should also be consistent. I'm +1 on LLVM style.

https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop

https://abseil.io/tips/227

E.g., the below snippet is an infinite loop.

```cpp
#include <iostream>
#include <vector>
using namespace std;

int main() {
  std::vector<int> x;
  for (int i = 0; i < x.size() - 1; ++i) {
    std::cout << "QQ" << std::endl;
  }
  return 0;
}
```

Reply:
It would be an infinite loop is unsigned indices were used, but thanks to the indices being signed here, it's actually a 0-iterations loop :-)

Reply:
Darn! Forgot that standard library containers had unsigned sizes. Point taken!

Reply:
Yes, the standard library uses unsigned sizes. :)

Reply:
Now I appreciate better the meaning of LLVM's style of initializing a second variable in the `for` for the loop bound. It's not about the optimization as the doc claims ("dont evaluate every time"). It's about insulating our loop condition from the ugly unsigned `size()` value. It doesn't matter anymore that `size()` returns unsigned as long as we only use the signed `e`.

Reply:
Exactly, this is what I feel when I investigate it today!